### PR TITLE
Adds more robust unit tests

### PR DESCRIPTION
**Pull request by bahmady**
Also adds method `#grid` to `::Grid` to make this work

### CODE REVIEW

File: lib/minesweeper/grid.rb [hunk 1]
```diff
@@ -46,5 +44,9 @@ def unseal(x, y)
     def dimensions
       [@x, @y]
     end
+
+    def grid
+      @grid
+    end
```
Comment (arnau) [inline lib/minesweeper/grid.rb]: There is a fundamental problem here. Why are you exposing the internals of the grid if the main purpose of having a class Grid is to abstract it away?

File: spec/grid_size_spec.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+require './lib/minesweeper'
+
+describe Minesweeper::Grid do
+  it "should produce an array of the right size" do
+    actual = Minesweeper::Grid.new(3, 3)
+    expect(actual.grid.map(&:count)).to eq([3, 3, 3])
```
Comment (arnau) [inline spec/grid_size_spec.rb]: This is testing the internals of the grid. Why is this important?

File: spec/grid_array_spec.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+require './lib/minesweeper'
+
+describe Minesweeper::Grid do
+  it "should produce an array" do
+    actual = Minesweeper::Grid.new(3,3)
+    expect(actual.grid).to be_instance_of(Array)
```
Comment (arnau) [inline spec/grid_array_spec.rb]: Why is this important?

File: lib/minesweeper/grid.rb (hunk 1 above)
Comment (arnau) [reply lib/minesweeper/grid.rb]: In short, the fact that the grid internally is an array should not matter to anyone, not even to your tests.

What matters is that the Grid has the right dimensions. Dimensions is a concept unrelated to how did you implement the grid.

File: spec/grid_size_spec.rb (hunk 2 above)
Comment (bahmady) [reply spec/grid_size_spec.rb]: So... I'm struggling to see how I can do something more robust than just pick out `::Grid.@x` and `::Grid.y`. The only way I can think of is to look at it in terms of the size of the array, for example `#count` the number of elements in the array. That's why I ended up doing it this way.

I can see how that's wrong, in hindsight, but what I can't see is how to do it: the only thing I can think of is doing something like `#instance_variable_get` on `::Grid.@grid` but that's the same as exposing it, which as you say in your other comment isn't the right approach.

So basically I think I'm stuck, and need a hint 😅

File: spec/grid_array_spec.rb (hunk 3 above)
Comment (bahmady) [reply spec/grid_array_spec.rb]: For the record I can totally see in hindsight why this isn't important. I just got carried away again. 

It's just what I tried to do in `spec/grid_size_spec.rb` that I'm stuck on.

File: spec/grid_size_spec.rb (hunk 2 above)
Comment (arnau) [reply spec/grid_size_spec.rb]: Ok, the _idea_ you implemented here is in the right direction, the implementation though is in the wrong place.

The task is to test the dimensions of the grid. The first test we did yesterday had the right interface:

```ruby
expect(actual.dimensions).to eq([3, 3])
```

So, that doesn't need to change at all. This means that your `Grid` has to have a `#dimensions` method that returns an Array of length 2 (because the grid is 2D) where the first element is the number of rows and the second element is the number of columns.

The original implementation didn't give you that, it gave you the parameters you passed to the constructor which doesn't guarantee the actual grid uses them at all.

The implementation you have in this test is in the right direction because you are getting the numbers from the grid. But you are not implementing a method! you are throwing the implementation out of your class for some reason. Also you are providing 3 values as a result which is not correct, we are dealing with a 2D grid, not a 3D cube.

Comment (arnau) [reply spec/grid_size_spec.rb]: Btw, you do not need any metaprogramming. I advice you not to go fancy here, you need a regular method `#dimensions`. The fundamental part is to remember that you must not cross the boundaries of your class.

**Comment by bahmady** [comment]
先生. 申し訳ありません.

File: lib/minesweeper/grid.rb [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
+
   class Grid
     def initialize(x, y)
-      @x = x
-      @y = y
+      @x, @y = x, y
       @grid = []
 
-      (0...x).each do ||
+      (0...@x).each do ||
```
Comment (arnau) [inline lib/minesweeper/grid.rb]: use a local parameter like before, there is no need to use an instance variable here.
